### PR TITLE
[mid_registrar] Fixes expires parameter when it must be 0.

### LINKED ISSUE #2722: [BUG] mid_registrar wrong expiry value
**OpenSIPS version you are running**
```
version: opensips 3.2.3 (x86_64/linux)
flags: STATS: On, DISABLE_NAGLE, USE_MCAST, SHM_MMAP, PKG_MALLOC, Q_MALLOC, F_MALLOC, HP_MALLOC, DBG_MALLOC, FAST_LOCK-ADAPTIVE_WAIT
ADAPTIVE_WAIT_LOOPS=1024, MAX_RECV_BUFFER_SIZE 262144, MAX_LISTEN 16, MAX_URI_SIZE 1024, BUF_SIZE 65535
poll method support: poll, epoll, sigio_rt, select.
git revision: f32089b8c
main.c compiled on  with gcc 8
```

**Describe the bug**
During unregister, "expires" parameter in Contact header is wrong.

**To Reproduce**
Example:
1. Start OpenSIPS
2. Use mid_registrar in mode "1"
3. Check REGISTER message sent to registrar when UA tries to un-register

**Expected behavior**
Expires parameter in Contact header must be "0".

**Relevant System Logs**
```
Nov 29 14:52:46 midregistrar1 /usr/sbin/opensips[26329]: DBG:mid_registrar:calc_ob_contact_expires: outgoing expires: 0
Nov 29 14:52:46 midregistrar1 /usr/sbin/opensips[26329]: DBG:mid_registrar:save_restore_rpl_contacts:     >> REGISTER -1638193966s ------- -1638193966s 200 OK <<!
Nov 29 14:52:46 midregistrar1 /usr/sbin/opensips[26329]: DBG:usrloc:get_ucontact: using ct matching mode 0
```

**OS/environment information**
 - Operating System: `Debian 10.10`
 - OpenSIPS installation: `git`

**Additional context**

### PR DESCRIPTION
**Pull request by vasilevalex**
Closes #2722

**Summary**
This check was in 2 places, but must be in 3.

### CODE REVIEW

File: modules/mid_registrar/save.c [hunk 1]
```diff
@@ -539,10 +539,7 @@ void overwrite_contact_expirations(struct sip_msg *req, struct mid_reg_info *mri
 	for (c = get_first_contact(req); c; c = get_next_contact(c)) {
 		calc_contact_expires(req, c->expires, &e, 1);
 		calc_ob_contact_expires(req, c->expires, &expiry_tick, mri->expires_out);
-		if (expiry_tick == 0)
-			new_expires = 0;
-		else
-			new_expires = expiry_tick - get_act_time();
+		new_expires = expiry_tick == 0 ? 0 : expiry_tick - get_act_time();
```
Comment (liviuchircu) [inline modules/mid_registrar/save.c:542]: Good idea to normalize the coding style (I approve!), although it is unrelated to the topic of the PR.  On small PRs, it is OK to do one or two of these, however on large PRs, having lots of such changes will make the reviewing more difficult.

File: modules/mid_registrar/save.c [hunk 2]
```diff
@@ -1348,7 +1345,7 @@ static inline int save_restore_rpl_contacts(struct sip_msg *req,
 			goto update_usrloc;
 
 		calc_ob_contact_expires(rpl, _c->expires, &e_out, 0);
-		e_out -= get_act_time();
+		e_out = e_out == 0 ? 0 : e_out - get_act_time();
```
Comment (liviuchircu) [inline modules/mid_registrar/save.c:1348]: The original code seems to have been written with the assumption that a 200 OK from the main registrar **cannot ever** include a Contact with **0** expiration (if a contact is included, it MUST have a non-zero expiry).  While it is a reasonable assumption to make, it makes the code very fragile to buggy or malicious SIP traffic, so your fix is excellent.

**Review by liviuchircu — APPROVED** [review]
Hey @vasilevalex and thank you for the fix!  Just to be sure that I correctly understood the bug, can you confirm that the 200 OK reply from the main-registrar includes a strange (possibly buggy) contact with a **0** expiration?

If you could include a "secured" version of a 200 OK reply that triggers the bug, it would be perfect!  Merging to `master` immediately, will backport as soon as I get feedback from you.

File: modules/mid_registrar/save.c (hunk 1 above)
Comment (vasilevalex) [reply modules/mid_registrar/save.c:542]: yes, I should make 2 commits.

File: modules/mid_registrar/save.c (hunk 2 above)
Comment (vasilevalex) [reply modules/mid_registrar/save.c:1348]: This was REGISTER from UA (to unregister):
```
REGISTER sip:test.voip.tel SIP/2.0
Via: SIP/2.0/UDP 52.10.16.16:64254;rport;branch=z9hG4bKPjEh7mH2LlAI5APPECJvR03S2YSqAYTNKi
Max-Forwards: 69
From: "new_test" <sip:102_test@test.voip.tel>;tag=1ZFs-hb.820fBtFb3b37tiygGdaIv2Qo
To: "new_test" <sip:102_test@test.voip.tel>
Call-ID: GHTV.sGO8eD1GrapKMKGAVna3fyROY2d
CSeq: 24771 REGISTER
User-Agent: Telephone 1.5.2
Contact: <sip:102_test@52.10.16.16:64254;ob>
Expires: 0
Content-Length:  0
```
Then main registrar sent 200 OK (its' behavior can be wrong, as it contains Contact with `expires=0`):
```
SIP/2.0 200 OK
Record-Route: <sip:172.18.16.241;r2=on;lr=on>,...
User-Agent: VoIP
Contact: <sip:102_test@172.26.26.231:5060;ctid=2356649242509991926;realip=52.10.16.16:64254;natip=52.10.16.16:64254>;expires=0
CSeq: 24771 REGISTER
Call-ID: GHTV.sGO8eD1GrapKMKGAVna3fyROY2d
Expires: 0
To: "new_test" <sip:102_test@test.voip.tel>
From: "new_test" <sip:102_test@test.voip.tel>;tag=1ZFs-hb.820fBtFb3b37tiygGdaIv2Qo
Content-Length: 0
Via: SIP/2.0/UDP 172.16.16.231:5060;branch=z9hG4bK904a.c3614d1.0
Via: SIP/2.0/UDP 52.10.16.16:64254;received=52.10.16.16;rport=64254;branch=z9hG4bKPjEh7mH2LlAI5APPECJvR03S2YSqAYTNKi
```
So mid_registrar transforms this reply to:
```
SIP/2.0 200 OK
Record-Route: <sip:172.18.16.241;r2=on;lr=on>,...
User-Agent: VoIP
CSeq: 24771 REGISTER
Call-ID: GHTV.sGO8eD1GrapKMKGAVna3fyROY2d
X-ID: hx-d82d416981dfd86c353434847c223082
To: "new_test" <sip:102_test@test.voip.tel>
From: "new_test" <sip:102_test@test.voip.tel>;tag=1ZFs-hb.820fBtFb3b37tiygGdaIv2Qo
Content-Length: 0
Via: SIP/2.0/UDP 52.10.16.16:64254;received=52.10.16.16;rport=64254;branch=z9hG4bKPjEh7mH2LlAI5APPECJvR03S2YSqAYTNKi
Contact: <sip:102_test@52.10.16.16:64254;ob>;expires=18446744072071357650;received="sip:52.10.16.16:64254"
```

Comment (liviuchircu) [reply modules/mid_registrar/save.c:1348]: Perfect!  Exactly as I imagined: the main-registrar replies with a 200 OK advertising a Contact with `;expires=0`:

* does this make sense?  Not really, as a `0` expiry means "this contact is valid for 0 more seconds".  So why advertise something that does not exist?
* is this a common occurrence?  Probably, and the mid-registrar implementation should be equipped to deal with any corner-cases on the downstream device without reaching a bad internal state.

Thank you for all your help, I can safely backport to `3.1`+ now.  And we are done here! :-)

**Pull request merged**